x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/TextEmbeddingResults.java

```diff
@@ -103,24 +105,60 @@ public Map<String, Object> asMap() {
         return map;
     }
 
-    public record Embedding(List<Float> values) implements Writeable, ToXContentObject {
+    public static class Embedding implements Writeable, ToXContentObject {
         public static final String EMBEDDING = "embedding";
 
+        public static Embedding ofFloats(List<Float> values) {
+            return new Embedding(convertFloatsToEmbeddingValues(values));
+        }
+
+        public static Embedding ofBytes(List<Byte> values) {
+            List<EmbeddingValue> convertedValues = values.stream().map(ByteValue::new).collect(Collectors.toList());
+
+            return new Embedding(convertedValues);
+        }
+
+        private final List<EmbeddingValue> values;
+
+        public Embedding(List<EmbeddingValue> values) {
+            this.values = values;
+        }
+
         public Embedding(StreamInput in) throws IOException {
-            this(in.readCollectionAsImmutableList(StreamInput::readFloat));
+            if (in.getTransportVersion().onOrAfter(TransportVersions.ML_INFERENCE_COHERE_EMBEDDINGS_ADDED)) {
```

Comment:
Double check that this is correct. This was added so we can support returning `bytes` or `floats`. I thought the best way to do that was to used a named writable but happy to change it to something else if there's a better solution. The writables that we support now are the `FloatValue` and `ByteValue` classes above.

Reply:
I could be completely wrong but this was my thinking 😅 : 

In this constructor `public Embedding(StreamInput in)`, I don't know if I need to read floats or bytes (since we're allowing either to occur now with the addition of cohere). So I think the way to "know" is to use a named writable which led me to add the `EmbeddingValue` and the `ByteValue`, `FloatValue`. That way when the results get serialized and another node needs to read them, it can determine whether to read them as floats or bytes.

Reply:
> rather than directly using primitives or Numbers

It does seem excessive to hold an array of EmbeddingValue objects when a primitive would do the job. `TextEmbeddingResults` is a named writable one option is create the `TextEmbeddingByteResults` and `ByteEmbeddings` classes for the byte variant. That would avoid the conditionals in the streaming functions.